resources/assets/coffee/_classes/store-supporter-tag.coffee

```diff
@@ -0,0 +1,150 @@
+###
+#    Copyright 2015-2017 ppy Pty. Ltd.
+#
+#    This file is part of osu!web. osu!web is distributed with the hope of
+#    attracting more community contributions to the core ecosystem of osu!.
+#
+#    osu!web is free software: you can redistribute it and/or modify
+#    it under the terms of the Affero GNU General Public License version 3
+#    as published by the Free Software Foundation.
+#
+#    osu!web is distributed WITHOUT ANY WARRANTY; without even the implied
+#    warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#    See the GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with osu!web.  If not, see <http://www.gnu.org/licenses/>.
+###
+
+class @StoreSupporterTag
+  RESOLUTION: 8
+  MIN_VALUE: 4
+  MAX_VALUE: 52
+
+  constructor: ->
+    $(document).on 'turbolinks:load', =>
+      @initialize(document.getElementById('js-store-supporter-tag'))
+    @debouncedGetUser = _.debounce @getUser, 300
+
+  initialize: (rootElement) =>
+    return unless rootElement?
+    @el = rootElement
+    @searching = false
+    @searchData = null
+    @currentUser =
+      username: @el.dataset.username
+      avatar_url: @el.dataset.avatarUrl
+
+    # Everything should be scoped under the root @el
+    @priceElement = @el.querySelector('.js-price')
+    @durationElement = @el.querySelector('.js-duration')
+    @discountElement = @el.querySelector('.js-discount')
+    @slider = @el.querySelector('.js-slider')
+    @sliderPresets = @el.querySelectorAll('.js-slider-preset')
+    @usernameInput = @el.querySelector('.js-username-input')
+
+    @initializeSlider()
+    @initializeSliderPresets()
+    @initializeUsernameInput()
+    @
+
+  initializeSlider: =>
+    slider = $(@slider).slider {
+      range: 'min',
+      value: @sliderValue(@MIN_VALUE),
+      min: @sliderValue(@MIN_VALUE),
+      max: @sliderValue(@MAX_VALUE),
+      step: 1,
+      animate: true,
+      slide: @onSliderValueChanged,
+      change: @onSliderValueChanged,
+    }
+    @updatePrice(@calculate(@sliderValue(@MIN_VALUE)))
+    slider
+
+  initializeSliderPresets: =>
+    $(@sliderPresets).on 'click', (event) =>
+      target = event.currentTarget
+      price = StoreSupporterTagPrice.durationToPrice(target.dataset.months)
+      $(@slider).slider('value', @sliderValue(price)) if price
+
+  initializeUsernameInput: =>
+    $(@usernameInput).on 'input', (event) =>
+      @onInput(event)
+
+  getUser: (username) =>
+    $.post laroute.route('users.check-username-exists'), username: username
+    .done (data) =>
+      @setUserInteraction(data?)
```

Comment:
model/visual behaviour separation; also, not enough Proxy or Observable support available :(

also, javascript :D